File: Robot Control Course/Chapter12/Lesson4/Chapter12_Lesson4.m

A = [0 1; 0 -2];
B = [0; 1];

h_nom = 0.002;
jitter_max = 0.0005;

Kp = 50;
Kd = 5;

x = [0.5; 0.0];
xref = 0.0;

T = 2.0;          % total simulated time
t = 0.0;
k = 0;
log_t = [];
log_x = [];

while t < T
    pos = x(1);
    vel = x(2);
    e = xref - pos;
    u = Kp * e - Kd * vel;

    h = h_nom + (2*rand - 1)*jitter_max;
    xdot = A * x + B * u;
    x = x + h * xdot;

    t = t + h;
    k = k + 1;
    log_t(k,1) = t;
    log_x(k,1:2) = x.';
end

plot(log_t, log_x(:,1));
xlabel('time (s)'); ylabel('position');
title('Joint position under sampling jitter');

File: Robot_Control_Course/Chapter12/Lesson4/Chapter12_Lesson4.py

import time
import math
import random
import numpy as np

# Simple scalar joint model: x = [position; velocity]
A = np.array([[0.0, 1.0],
              [0.0, -2.0]])   # some damping
B = np.array([[0.0],
              [1.0]])         # torque input

# Nominal sampling period (seconds) and jitter bound
h_nom = 0.002   # 2 ms
jitter_max = 0.0005  # +/- 0.5 ms

# PD gains
Kp = 50.0
Kd = 5.0

x = np.array([[0.5],
              [0.0]])  # initial position error
xref = 0.0

def step_euler(x, u, h):
    return x + h * (A @ x + B * u)

t = 0.0
for k in range(2000):
    # Compute control
    pos = x[0, 0]
    vel = x[1, 0]
    e = xref - pos
    u = Kp * e - Kd * vel

    # Draw jittered step size
    h = h_nom + random.uniform(-jitter_max, jitter_max)

    # Integrate
    x = step_euler(x, u, h)
    t += h

    if k % 100 == 0:
        print(f"k = {k}, t = {t:.4f}, pos = {pos:.4f}, h = {h:.6f}")

    # In a real RT system the sleep is driven by a timer interrupt
    time.sleep(h_nom)
